### ISSUE TITLE
using `pmap` causing slowly gpu memory increases and eventually `OOM` in the end 

### ISSUE THREAD

**Issue by WingsOfPanda** [jax-ml/jax#11357]
Dear team:

I am trying to use `pmap` to train a simple CNN network across multiple GPUs, and, as title says, as training goes on, the gpu memory consumption is slowly increasing and finally causing `OOM` issue. 

Let me use the following screen shot to better explain the issues.

We use four 3090 GPUs for demonstration. As we just start training, we can see from the below screen shot that `GPU0` is using 23379MB memory and `GPU1`-`GPU3` are using 23355MB memory. 
![Screen Shot 2022-07-03 at 9 02 29 PM](https://user-images.githubusercontent.com/46889518/177040998-140458fa-1ced-4e0d-b143-e948b0fb0332.png)

After we train for couple of steps, we can see from the below screen shot that`GPU0` are using 23419MB memory and `GPU1`-`GPU3` are using 23395MB memory. Almost 40MB memory increase on each GPU. 
![Screen Shot 2022-07-03 at 9 05 44 PM](https://user-images.githubusercontent.com/46889518/177041096-495f1c36-f48e-4658-909f-dd0e889466fa.png)

In the end, we shall reach the `OOM` situation...

The following code is a minimal working example that could reproduce the error. I am using four 24GB 3090 GPUs. 

```
import os, sys
import time

from functools import reduce, partial
from collections import deque
from typing import *

import numpy as np

import jax
import jax.numpy as jnp
from jaxlib import xla_client as xc

import optax

import flax
from flax.training import train_state, checkpoints
from flax import linen as nn


class SimpleConvLayer(nn.Module):
    feature: int = 4
    kernel: Tuple[int] = (3, 3, 3)
    stride: int = 1

    def setup(self):
        self.conv = nn.Conv(self.feature, self.kernel, self.stride)
        self.norm = nn.GroupNorm(num_groups=None, group_size=4)

    def __call__(self, x):
        x = self.conv(x)
        x = self.norm(x)
        x = nn.relu(x)
        return x


class YOLOV1(nn.Module):

    @nn.compact
    def __call__(self, x):
        x = SimpleConvLayer(feature=64, kernel=(7, 7), stride=2)(x)
        x = nn.max_pool(x, (2, 2), (2, 2))

        x = SimpleConvLayer(feature=192, kernel=(3, 3))(x)
        x = nn.max_pool(x, (2, 2), (2, 2))

        x = SimpleConvLayer(feature=128, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=256, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=256, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=512, kernel=(3, 3))(x)
        x = nn.max_pool(x, (2, 2), (2, 2))

        x = SimpleConvLayer(feature=256, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=512, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=256, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=512, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=256, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=512, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=256, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=512, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=512, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=1024, kernel=(3, 3))(x)
        x = nn.max_pool(x, (2, 2), (2, 2))

        x = SimpleConvLayer(feature=512, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=1024, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=512, kernel=(1, 1))(x)
        x = SimpleConvLayer(feature=1024, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=1024, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=1024, kernel=(3, 3), stride=2)(x)

        x = SimpleConvLayer(feature=1024, kernel=(3, 3))(x)
        x = SimpleConvLayer(feature=1024, kernel=(3, 3))(x)

        x = jnp.reshape(x, (-1, x.shape[1] * x.shape[2] * x.shape[3]))
        x = nn.Dense(4096)(x)
        x = nn.relu(x)
        x = nn.Dense(4410)(x)
        x = nn.sigmoid(x)
        x = jnp.reshape(x, (-1, 7, 7, 90))
        return x


def get_create_train_state(module: flax.linen.Module, devices=None):
    def create_train_state(key, input_shape, lr=0.001):
        # 1. Model instance
        model = module()

        # 2. Initialize the parameters of the model
        params = model.init(key, jnp.ones(input_shape))['params']

        # 3. Define the optimizer with the desired learning rate
        optimizer = optax.adam(learning_rate=lr)

        return train_state.TrainState.create(apply_fn=model.apply, params=params, tx=optimizer)

    return jax.pmap(create_train_state, axis_name="devices", in_axes=0, out_axes=0, static_broadcasted_argnums=(1, 2),
                    devices=devices)


def get_calculate_loss(module: nn.Module,
                       parallel_devices: Optional[List[xc.Device]] = None,
                       loss_device: xc.Device = jax.devices('cpu')[-1]):
    def calculate_loss(params, xx, yy):
        # test whether there are some device placement requirement

        def forward_pass(params, input_data):
            prediction = module().apply({"params": params}, input_data)
            # print(f'-*- prediction shape: {prediction.shape}')
            return prediction

        pmap_func = jax.pmap(
            forward_pass,
            static_broadcasted_argnums=(),  # broadcast the paramss
            axis_name="device",
            in_axes=0,
            out_axes=0,
            devices=parallel_devices,
            donate_argnums=()
        )
        # be aware that all outputs are sharded array, then should be collect to a single device to calculate loss
        prd_ = pmap_func(params, xx)

        def forward_loss(lab, prd):
            send_to_device = partial(jax.device_put, device=loss_device)

            lab = jax.tree_map(send_to_device, lab)
            prd = jax.tree_map(send_to_device, prd)

            # then remember to fetch the data to same device

            loss = jnp.mean(jnp.square(lab - prd))
            return loss

        # return jax.jit(forward_loss)(yy, prd_)
        return forward_loss(yy, prd_)

    return calculate_loss


def get_apply_model(calculate_loss):

    def apply_model(state, X, Y):
        loss, grads = jax.value_and_grad(calculate_loss, has_aux=False)(state.params, X, Y)

        return grads, loss

    # return jax.jit(apply_model)
    return apply_model


def get_update_model(devices: List[xc.Device]):

    axis_name = "devices"  # a name for the axis 0

    def update_model(state, grads):
        grads = jax.lax.pmean(grads, axis_name=axis_name)
        return state.apply_gradients(grads=grads)  # apply two grads to states?

    return jax.pmap(update_model, axis_name=axis_name, in_axes=0, out_axes=0, devices=devices)


def try_get_device(device_type: str) -> List[xc.Device]:
    try:
        return jax.devices(device_type)
    except RuntimeError as re:
        sys.stderr.write("No device for device type: \"{}\"\n".format(device_type))
        return []


def fake_demostration() -> None:
    random_key = jax.random.PRNGKey(0)
    random_key, spl = jax.random.split(random_key)

    cpu_devices, gpu_devices = try_get_device("cpu"), try_get_device("gpu")

    forward_device = gpu_devices if gpu_devices else cpu_devices
    weight_device = cpu_devices[-1]

    spl = jnp.array([spl])
    spl = jnp.tile(spl, (len(forward_device), 1))

    batch_size = 8
    effective_batch_size = batch_size * len(forward_device)  # effective batch size = batch_size * device count
    epochs = 256
    image_size = (448, 448, 1)  # for update_model
    grid_size = (7, 7, 90)

    calculate_loss = get_calculate_loss(module=YOLOV1, parallel_devices=forward_device, loss_device=weight_device)
    train_state_ = get_create_train_state(YOLOV1, devices=forward_device)(spl, (batch_size, *image_size), 1e-4)
    apply_model = get_apply_model(calculate_loss)
    update_model = get_update_model(devices=forward_device)

    loss_queue = deque(maxlen=10)
    steps_per_epoch = 0
    step_count = True
    for epoch_num in range(epochs):
        step_num = 0
        while True:

            dts = time.perf_counter_ns()
            X0, X1 = np.random.randn(effective_batch_size, *image_size), np.random.randn(effective_batch_size, *grid_size)
            dte = time.perf_counter_ns()

            t0 = time.perf_counter_ns()

            X0 = [jax.device_put(x) for x in X0.reshape((len(forward_device), batch_size, *image_size))]
            X1 = [jax.device_put(x) for x in X1.reshape((len(forward_device), batch_size, *grid_size))]

            X0 = jax.device_put_sharded(X0, devices=forward_device)
            X1 = jax.device_put_sharded(X1, devices=forward_device)

            grads, loss = apply_model(train_state_, X0, X1)
            train_state_ = update_model(train_state_, grads)

            loss_queue.append(loss)
            t1 = time.perf_counter_ns()
            step_num += 1
            steps_per_epoch += 1 if step_count else 0
            print("[{3:4d}] {0:4d}/{1:4d} loss={2:.4e}, mean loss = {4:.4e}, data time = {5:.3e} s, ops time = {6:.3e} s".format(
                step_num + 1,
                steps_per_epoch,
                loss,
                epoch_num + 1,
                sum(loss_queue) / len(loss_queue),
                (dte - dts) * 1e-9,
                (t1 - t0) * 1e-9), end="\r")

    sys.exit(0)


if __name__ == "__main__":
    fake_demostration()

```

**Comment by hawkinsp** [comment]
A couple of quick questions:
* what `jax` and `jaxlib` versions are you using?
* have you seen this on any other GPUs?

**Comment by WingsOfPanda** [comment]
@hawkinsp hi, I am using `jax==0.3.14` and `jaxlib==0.3.14`. I only tested this on 2080 and 3090 GPU, both of them experiencing the same issue

**Comment by pschuh** [comment]
Thanks for the report. This is due to repeated compilation inside calculate_loss() (if you lift pmap_func outside, the leak will go away and is probably what you want). Also, as a side note, you can just pmean your loss and then pmap the entire train step (rather than running value_and_grad over the forward pmap). It has certainly been interesting trying to track down the leak in the repeated compilation process, but no luck yet.
